tests/integration_tests/data_steward/cdr_cleaner/cleaning_rules/cleaning_tests_base.py

```diff
@@ -0,0 +1,229 @@
+"""
+Integration Test base classes for cleaning rules.
+
+Implements a base class that will ensure test datasets and empty tables with
+schemas are created before beginning the test.  It will remove tables and test
+datasets at cleanup.  It will provide optional classes that extend this
+base class for simple rules that only modify data or drop rows.  These special
+classes can be extended to use their test execution parameters.  If a more
+detailed implementation is needed, the developer can still extend the base class
+only and implement their own tests.
+"""
+# Python imports
+import unittest
+
+# Third party imports
+import google.cloud.exceptions as gc_exc
+
+# Project imports
+from cdr_cleaner import clean_cdr_engine as engine
+from utils import bq
+
+
+class BaseTest:
+
+    class CleaningRulesTestBase(unittest.TestCase):
+        """
+        A base class for cleaning rules integration tests.
+
+        This class implements setUp, setUpClass, tearDown, and tearDownClass
+        methods that can be inherited and used by all extending classes.  This
+        allows developers to focus more fully on creating a working test rather
+        than setting up or tearing down their test environment.
+
+        The base class ensures test and sandbox datasets are created.  It also
+        creates empty test tables with correct schemas to further validate the
+        cleaning rule sql.  The base class ensures the test tables are reloaded
+        with the sample data between tests.  It also ensures tables are deleted
+        when the test is finished.
+        """
+
+        # this should always be for a test project.  a check is implemented to
+        # make sure this is true
+        project_id = ''
+        # a list of fully qualified table names the cleaning rule is targeting.
+        # fq = {project_id}.{dataset_id}.{table_name}.  required for
+        # test setup and cleanup.
+        fq_table_names = []
```

Comment:
It feels like there may be a layer here, or possibly this whole base test (minus `drop_rows`), which pertains to BigQuery testing in general (not just cleaning) and which we could potentially use elsewhere, e.g. retraction

Reply:
yes.  that is the point.  The most basic class simply deals with setting up the required datasets and empty tables with correct schema definitions. The extending base class offers a simple way to test a class who's intention is to drop rows based on some criterion.  tests that don't drop rows shouldn't extend that particular class. 

`drop_rows()` is still needed to clear the tables between each test without dropping the entire table.  the entire table is dropped when the test class is finished.

Reply:
I changed the module name and the base class name to try to make it more readily usable by other tests, i.e. retraction.  I think it messed up the comments here.